File: singleton_pattern.go
package main

import (
	"fmt"
	"sync"
	"time"

	"github.com/sirupsen/logrus"
)

// Simulates resource-intensive logger creation
func New() *logrus.Logger {
	time.Sleep(1 * time.Second) // resource initialization -> network connections, file connections, etc
	return logrus.New()
}

// Function to create a new logger instance every time
func GetNewLogger() *logrus.Logger {
	return New()
}

// Singleton pattern ensures that only one instance of the logger is created
var (
	loggerOnce sync.Once
	logger     *logrus.Logger
)

func GetSingletonLogger() *logrus.Logger {
	loggerOnce.Do(func() {
		logger = New()
	})
	return logger
}

func main() {
	fmt.Println("Starting application...")

	// Example 1: New logger instance every time
	fmt.Println("Using a new logger instance each time:")
	for i := 0; i < 5; i++ {
		log := GetNewLogger()
		log.Infof("Log message %d from new logger", i+1)
	}

	// Example 2: Singleton logger
	fmt.Println("Using a singleton logger instance:")
	for i := 0; i < 5; i++ {
		log := GetSingletonLogger()
		log.Infof("Log message %d from singleton logger", i+1)
	}
}

File: java/singleton_pattern/SingletonPattern.java
import java.util.logging.Logger;

public class SingletonPattern {

    private static Logger newLogger() {
        try {
            Thread.sleep(1000);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
        return Logger.getLogger("app-" + System.nanoTime());
    }

    public static Logger getNewLogger() {
        return newLogger();
    }

    private static class Holder {
        private static final Logger INSTANCE = newLogger();
    }

    public static Logger getSingletonLogger() {
        return Holder.INSTANCE;
    }

    public static void main(String[] args) {
        System.out.println("Starting application...");

        System.out.println("Using a new logger instance each time:");
        for (int i = 0; i < 5; i++) {
            Logger log = getNewLogger();
            log.info("Log message " + (i + 1) + " from new logger");
        }

        System.out.println("Using a singleton logger instance:");
        for (int i = 0; i < 5; i++) {
            Logger log = getSingletonLogger();
            log.info("Log message " + (i + 1) + " from singleton logger");
        }
    }
}
